### DISCUSSION TITLE
HA-Yellow: Nextcloud and its own nvme: proof of concept

### DISCUSSION THREAD

**Discussion by senna1992** [alexbelgium/hassio-addons#1370] [Q&A] [answered]
Hi there,

this is my journey:

the goal? i want to host my cloud myself.
the basics? homeassistant yellow with m.2 ssd and cm4 with emmc.

the obstacles?
running nextcloud ootb it stores everything inside of the emmc - and therefore bloats the HASS backups.
running HASS on the same partition as nextclouds data storage - a crash could cuase huge data loss.
every upload runs through the emmc ootb, so it crashed my smarthome for about 2 days repeatedly.

my result:
installed the m.2 nvme preformatted to ext4.
installed nextcloud on the pi and added the ssd to the addons config under local disc.
changed the temp directory used by nextcloud (using docker to access the containers console).

How to do it?
- First get a good grade ssd (i got western digital RED because they say it is rated for nas usage)
- make it ext4 and create two folders (data and temp)
- grab the /dev/name using "advanced ssh terminal" addon and command "lsblk -f"
- drink coffee
- setup nextcloud, mariadb (the nc protocol view shows what to enter as server for your mariadb)
- edit two config files (see below) within the container (/data/config/php/www2.conf and /data/config/www/nextcloud/config/config.php)
- run chown abc:abc /path/to/your/tempfolder
- run chmod u+w /path/to/your/tempfolder
- restart
- check folder size (growing) with du -h -s /path/to/your/tempfolder while uploading something using the pc/mac client.

What to add inthe mentioned files?
1) config.php: add 'tempdirectory' => '/path/to/your/tempfolder'**,**
2) www2.conf: add 

- env[TMP] = /path/to/your/tempfolder
- env[TMPDIR] = /path/to/your/tempfolder
- env[TEMP] = /path/to/your/tempfolder

now it should work and spare your precious emmc

**Comment by alexbelgium** [comment, marked as answer]
Great how-to thanks very much! I'll add a link from the readme page

**Comment by purierca** [comment]
Hi there @alexbelgium and @senna1992 thanks for sharing, as I'm looking to implement this I have a few questions:

Please validate my understanding of the process:
1. Get **SSD**, format to ext4, create two folders `data` and `temp`
2. Plug it in, add it as local disc in **NextCloud** add-on config
3. Obtain the `/dev/name using` using the **Advanced SSH Terminal add-on**, with the command `lsblk -f`
4. Start **Nextcloud**, using **MariaDB** as database (follow the usual doc for this)
5. Open the **Nextcloud** container using **Docker add-on** and change the following:
   - `/data/config/php/www2.conf `: add 'tempdirectory' => '/path/to/your/tempfolder'
   - `/data/config/www/nextcloud/config/config.php` :     add
      - env[TMP] = /path/to/your/tempfolder
      - env[TMPDIR] = /path/to/your/tempfolder
      - env[TEMP] = /path/to/your/tempfolder
8. Back in the **Advanced SSH Terminal add-on**:
   - Run `chown abc:abc /path/to/your/tempfolder`
   - Run `chmod u+w /path/to/your/tempfolder`
11. Restart HA instance
12. Bonus/validation: Check that the folder size is growing while uploading something using the desktop client, with command `du -h -s /path/to/your/tempfolder` in the **Advanced SSH Terminal add-on**.

Questions:
- Can I use the Docker add-on for a single operation, then delete it? I ask because I've seen that it removes the HA OS "official" status including updates. Can I perform the same steps without Docker?
- Does the format have to be ext4 or can I use (just for example) NTFS, maybe through the Samba NAS add-on?
- Can I make the change of an already setup Nextcloud add-on, or do I need a fresh install?

**Comment by senna1992** [comment]
Questions:
1) I suggest using "portainer" which is a supported addon, that allows you to get to the console of the NC container.'
2) It might, never tried. The partition is mounted to the container, as if it was inside the machine. no NAS-Addon here.
3)  you can, edit the existing setup, but please have a backup of the addon and your personal data first!!


Notes to the Step by Step guide:
 7) restart NextCloud addon, not HA.
 8) can also be done right from the dockers container console, seems more intuite (use portainer to get to the console)

**Comment by purierca** [comment]
@senna1992 thanks a bunch.
Where is the /dev/name used in the process? I ask because I can't get it from the Terminla & SSH:

![image](https://github.com/user-attachments/assets/d041d901-f0e1-4765-af17-b0e98e651c43)

**Comment by senna1992** [comment]
I found another way, which seems much more convenient (only applicable if you have the ability to preformat the disk to ext4):

Go to Settings, Hardware, complete / all hardware and then filter for nvme.
![2024-08-02 17 42 24 192 168 0 42 0533c10aab4a](https://github.com/user-attachments/assets/ae08a3a1-1815-4637-9523-7cb653fa4f95)


In my case the needed path is found here:
![2024-08-02 17 43 19 192 168 0 42 c546f0c55753](https://github.com/user-attachments/assets/a7e727d2-aa37-4cc3-b260-b86c83bb70f5)

**Reply by purierca** [reply]
Great thanks!
And then you use the dev name for the full path to the temp folder is that it's?

**Reply by senna1992** [reply]
![2024-08-02 18 02 25 192 168 0 42 78cab37df56a](https://github.com/user-attachments/assets/5f5afb8e-7cc7-4a5c-afd7-25070d8b0b0d)


This is how i use it (taken from addon settings)

**Comment by senna1992** [comment]
do avoid confusion, here are my config files in changed form: [access them using vim inside of the container's terminal with portainer on HA]
www2.conf:
![www2-conf](https://github.com/user-attachments/assets/c09db1ab-3414-4cb3-847e-b865f572e54f)

config.php
![config-php](https://github.com/user-attachments/assets/a0284764-23d4-4136-b485-c696975837a3)

**Comment by purierca** [comment]
@senna1992 sorry for the noob questions, I'm new to this; for the life of me I can't find the variables to update in Portainer. It's the Environment ones, right? I can't find one that doesn't look like a bad idea to edit...

![image](https://github.com/user-attachments/assets/1cc47635-d00e-4250-9592-908561e437f8)

**Reply by senna1992** [reply]
That's good, to ask before changing something wrong :)


Please start with some basic vim tutorials to get to know how to use it.

Vim will be the tool you need to modify the files.

First get to learn how to use vim,
(open a file, edit the content, save the file)

If you are a top tier newbie, get to know bash terminals first :) or just do a backup.

Once you have done those, the fun part starts:


- open portainer, select containers, select nextcloud, access the terminal from there
- use cd to get to the folders where the file in question is located. 
- use vim to edit and save the file
- restart nextcloud addon

**Reply by purierca** [reply]
Lots to learn! The steps you detail make a lot of sense thank you. I'll dig into this and come back only if I have terminal questions. Thanks for the help @senna1992

**Reply by aiq008** [reply]
Is the any way of doing this with homeassistant OS running on a rasberry pi? I am have trouble giving permissions to the external drive

**Comment by saken2** [comment]
> Hi there,
> 
> this is my journey:
> 
> the goal? i want to host my cloud myself. the basics? homeassistant yellow with m.2 ssd and cm4 with emmc.
> 
> the obstacles? running nextcloud ootb it stores everything inside of the emmc - and therefore bloats the HASS backups. running HASS on the same partition as nextclouds data storage - a crash could cuase huge data loss. every upload runs through the emmc ootb, so it crashed my smarthome for about 2 days repeatedly.
> 
> my result: installed the m.2 nvme preformatted to ext4. installed nextcloud on the pi and added the ssd to the addons config under local disc. changed the temp directory used by nextcloud (using docker to access the containers console).
> 
> How to do it?
> 
>     * First get a good grade ssd (i got western digital RED because they say it is rated for nas usage)
> 
>     * make it ext4 and create two folders (data and temp)
> 
>     * grab the /dev/name using "advanced ssh terminal" addon and command "lsblk -f"
> 
>     * drink coffee
> 
>     * setup nextcloud, mariadb (the nc protocol view shows what to enter as server for your mariadb)
> 
>     * edit two config files (see below) within the container (/data/config/php/www2.conf and /data/config/www/nextcloud/config/config.php)
> 
>     * run chown abc:abc /path/to/your/tempfolder
> 
>     * run chmod u+w /path/to/your/tempfolder
> 
>     * restart
> 
>     * check folder size (growing) with du -h -s /path/to/your/tempfolder while uploading something using the pc/mac client.
> 
> 
> What to add inthe mentioned files?
> 
>     1. config.php: add 'tempdirectory' => '/path/to/your/tempfolder'**,**
> 
>     2. www2.conf: add
> 
> 
>     * env[TMP] = /path/to/your/tempfolder
> 
>     * env[TMPDIR] = /path/to/your/tempfolder
> 
>     * env[TEMP] = /path/to/your/tempfolder
> 
> 
> now it should work and spare your precious emmc

Hello, Can I also use an external regular SSD within an USB enclosure ?

**Comment by senna1992** [comment]
Yes but make sure to use uuid or another unique permanent identifier

________________________________

**Reply by saken2** [reply]
> uuid
Hi, Thanks for your reply. I'm kinda a newbie in Linux. I installed HA with some addons on my pi5b (qbittorrent, Calibre, Firefox , Emby etc), and I am very excited.  Now for the firs time installing the Nextcloud addon. I don't know what UUID is. I have to find out.
